File: Data Structures & Algorithms/remove-node-from-end-of-linked-list/submission-6.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode* curr = head;
        int count=0;
        while(curr){
            curr=curr->next;
            count++;
        }
        ListNode* curr2=head;
        ListNode* prev = nullptr;
        int count2=0;
        if(count-n==0){
            return head->next;
        }
        while(curr2){
            if(count2==count-n){
                prev->next = curr2->next;
                break;
            }
            prev = curr2;
            curr2 = curr2->next;
            count2++;
        }
        return head;
    }
};

File: Data Structures & Algorithms/copy-linked-list-with-random-pointer/submission-6.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        temp = {}
        cur = head
        while cur:
            temp[cur] = Node(cur.val)
            cur = cur.next
        temp[None] = None
        curr = head
        while curr:
            temp[curr].next = temp[curr.next]
            temp[curr].random = temp[curr.random]
            curr = curr.next
        return temp[head]
        


    